stdlib/public/core/Substring.swift.gyb

```diff
@@ -362,24 +377,107 @@ extension StringProtocol {
+%   ('characters', 'Character')
+% ]:
+%   View = ViewPrefix + 'View'
+%   RangeReplaceable = 'RangeReplaceable' if property == 'unicodeScalars' else ''
+extension Substring {
+  public struct ${View} {
+    var _slice: ${RangeReplaceable}BidirectionalSlice<String.${View}>
+  }
+}
+
+extension Substring.${View} : BidirectionalCollection {
+  public typealias Index = String.Index
+  public typealias SubSequence = Substring.${View}
 
-  public var ${property}: String.${ViewPrefix}View {
+  /// Creates an instance that slices `base` at `_bounds`.
+  internal init(_ base: String.${View}, _bounds: Range<Index>) {
+    _slice = ${RangeReplaceable}BidirectionalSlice(
+      base: String(base._core).${property},
+      bounds: _bounds)
+  }
+
+  /// The entire String onto whose slice this view is a projection.
+  internal var _wholeString: String {
+    return String(_slice._base._core)
+  }
+
+  public var startIndex: Index { return _slice.startIndex }
+  public var endIndex: Index { return _slice.endIndex }
+  public func index(after i: Index) -> Index {
+    return _slice.index(after: i)
+  }
+  public func index(before i: Index) -> Index {
+    return _slice.index(before: i)
+  }
+  public subscript(i: Index) -> String.${View}.Element {
+    return _slice[i]
+  }
+  public subscript(r: Range<Index>) -> SubSequence {
+    return Substring.${View}(_wholeString.${property}, _bounds: r)
+  }
+}
+
+extension Substring {
+  public var ${property}: ${View} {
     get {
-      return String(self).${property}
+      return ${View}(_wholeString.${property}, _bounds: startIndex..<endIndex)
     }
     set {
-      let base = String(describing: newValue)
-      self = Substring(_base: base, base.startIndex ..< base.endIndex)
+      self = Substring(newValue)
     }
   }
+
+  /// Creates a Substring having the given content.
+  ///
+  /// - Complexity: O(1)
+  public init(_ content: ${View}) {
+    self = content._wholeString[content.startIndex..<content.endIndex]
+  }
+}
+
+extension String {
+  % if property in ('utf8', 'utf16'):
+  /// Creates a String having the given content.
+  ///
+  /// If `codeUnits` is an ill-formed code unit sequence, the result is `nil`.
+  ///
+  /// - Complexity: O(N), where N is the length of the resulting `String`'s
+  ///   UTF-16.
+  public init?(_ codeUnits: Substring.${View}) {
+    let wholeString = codeUnits._wholeString
+    guard
+      codeUnits.startIndex.samePosition(in: wholeString.unicodeScalars) != nil
+      && codeUnits.endIndex.samePosition(in: wholeString.unicodeScalars) != nil
+    else { return nil }
+    self = String(Substring(codeUnits))
+  }
+  % else:
+  /// Creates a String having the given content.
+  ///
+  /// - Complexity: O(N), where N is the length of the resulting `String`'s
+  ///   UTF-16.
+  public init(_ content: Substring.${View}) {
+    self = String(Substring(content))
+  }
+  % end
+}
 % end
+
+// FIXME: The other String views should be RangeReplaceable too.
```

Comment:
No, I don't know. @dabrahams do you recall why? Otherwise I can investigate later.